<!--
  This Behavior Tree enables a robot to return to the 
  charging dock when the battery is low.
-->

<root main_tree_to_execute="MainTree">
  <BehaviorTree ID="MainTree">
    <Fallback name="NavigateOrRecharge">
      <ReactiveSequence name="Navigate">
        <IsBatteryLow min_battery="0.25" battery_topic="/battery_status" is_voltage="false"/>          
        <RecoveryNode number_of_retries="6" name="NavigateRecovery">
          <PipelineSequence name="NavigateWithReplanning">
            <RateController hz="0.333">
              <RecoveryNode number_of_retries="1" name="ComputePathThroughPoses">
                <ReactiveSequence>
                  <RemovePassedGoals input_goals="{goals}" output_goals="{goals}" radius="0.7"/>
                  <ComputePathThroughPoses goals="{goals}" path="{path}" planner_id="GridBased"/>
                </ReactiveSequence>
                <ReactiveFallback name="ComputePathThroughPosesRecoveryFallback">
                  <GoalUpdated/>
                  <ClearEntireCostmap name="ClearGlobalCostmap-Context" service_name="global_costmap/clear_entirely_global_costmap"/>
                </ReactiveFallback>
              </RecoveryNode>
            </RateController>
            <RecoveryNode number_of_retries="1" name="FollowPath">
              <FollowPath path="{path}" controller_id="FollowPath"/>
              <ReactiveFallback name="FollowPathRecoveryFallback">
                <GoalUpdated/>
                <ClearEntireCostmap name="ClearLocalCostmap-Context" service_name="local_costmap/clear_entirely_local_costmap"/>
              </ReactiveFallback>
            </RecoveryNode>
          </PipelineSequence>
          <ReactiveFallback name="RecoveryFallback">
            <GoalUpdated/>
            <RoundRobin name="RecoveryActions">
              <Sequence name="ClearingActions">
                <ClearEntireCostmap name="ClearLocalCostmap-Subtree" service_name="local_costmap/clear_entirely_local_costmap"/>
                <ClearEntireCostmap name="ClearGlobalCostmap-Subtree" service_name="global_costmap/clear_entirely_global_costmap"/>
              </Sequence>
              <Spin spin_dist="1.57"/>
              <Wait wait_duration="5"/>
              <BackUp backup_dist="0.15" backup_speed="0.025"/>
            </RoundRobin>
          </ReactiveFallback>
        </RecoveryNode> 
      </ReactiveSequence> 
      <Spin spin_dist="-3.14"/>
    </Fallback>  
  </BehaviorTree>
</root>
